src/unix/filesystem.c

```diff
@@ -0,0 +1,150 @@
+#include <unix_internal.h>
+#include <filesystem.h>
+
+#define SYMLINK_HOPS_MAX    8
+
+// fused buffer wrap, split, and resolve
+int resolve_cstring(tuple cwd, const char *f, tuple *entry, tuple *parent)
+{
+    if (!f)
+        return -EFAULT;
+
+    tuple t = *f == '/' ? filesystem_getroot(current->p->fs) : cwd;
+    tuple p = t;
+    buffer a = little_stack_buffer(NAME_MAX);
+    char y;
+    int nbytes;
+    int err;
+
+    while ((y = *f)) {
+        if (y == '/') {
+            if (buffer_length(a)) {
+                p = t;
+                t = lookup(t, intern(a));
+                if (!t)
+                {
+                    err = -ENOENT;
+                    goto done;
+                }
+                err = filesystem_follow_links(t, p, &t);
+                if (err) {
+                    t = false;
+                    goto done;
+                }
+                if (!children(t))
+                    return -ENOTDIR;
```

Comment:
Why are some of these direct returns as opposed to return via done?

Reply:
Direct returns are supposed to be used when the return code is different from -ENOENT, in which case the caller shouldn't bother checking what values have been put into *entry and *parent. (The only error case where *parent is set is for -ENOENT.) But now I see that there is a direct return -ENOENT in case of invalid UTF-8 sequence, will fix that.